gymnasium/utils/serialize_spec_stack.py

```diff
@@ -0,0 +1,123 @@
+"""Functions that use/modify specification stacks."""
+import dataclasses
+import inspect
+import json
+import re
+from typing import Union
+
+from gymnasium import error
+from gymnasium.dataclasses import WrapperSpec
+from gymnasium.envs.registration import EnvSpec
+
+
+def serialise_spec_stack(stack: tuple) -> str:
+    """Serialises the specification stack into a JSON string.
+
+    Args:
+        stack: Tuple of environment and wrapper specifications, known as the specification stack. Generated via `env.spec_stack`.
+
+    Returns:
+        A JSON string representing the specification stack.
+    """
+    num_layers = len(stack)
+    stack_json = {}
+    for i, spec in enumerate(stack):
+        spec = _serialise_callable(spec)
+        if i == num_layers - 1:
+            layer = "raw_env"
+        else:
+            layer = f"wrapper_{num_layers - i - 2}"
+        spec_json = json.dumps(dataclasses.asdict(spec))
+        stack_json[layer] = spec_json
+    return stack_json
+
+
+def _serialise_callable(spec: Union[WrapperSpec, EnvSpec]) -> dict:
+    """Converts a spec's lambda functions to a string representation.
+
+    Args:
+        spec: An environment or wrapper specification.
+
+    Returns: The specification with lambda functions converted to strings.
+
+    """
+    for k, v in spec.kwargs.items():
+        if callable(v):
+            try:
+                str_repr = (
+                    str(inspect.getsourcelines(v)[0]).strip("['\\n']").split(" = ")[1]
+                )  # https://stackoverflow.com/a/30984012
+            except OSError:
+                raise error.Error(
+                    "The attempted seialisation of a callable failed. This is likely due to env.spec_stack being called twice, which for technical reasons doesn't work. Please modify your code to only call env.spec_stack once (you can save that to a variable and repeatedly use that)."
+                )
+            str_repr = re.search(r", (.*)\)$", str_repr).group(1)
```

Comment:
Raises the precommit error " error: "group" is not a known member of "None" (reportOptionalMemberAccess"

but idk why

Reply:
I suspect that it is because, `re.search` has an optional return type. Therefore you just need to split up the statement if to two, if a check in the middle that the result of `search` is not None